File: RA 2.1/Actividad 5.py
class Empleado: 
    def __init__(self, nombre, salario_base):
        self.nombre= nombre
        self.salario_base= salario_base

    def calcular_salario(self): 
        return self.salario_base
    
    def __add__(self, otro): 
        if isinstance(otro, Empleado): 
            return self.calcular_salario() + otro.calcular_salario()
        return NotImplemented
    
    def __lt__(self, otro):
        if isinstance(otro, Empleado): 
            
        return NotImplemented

    def __str__(self): 
        return f"{self.nombre}: ${self.calcular_salario():.2f}"
    
class EmpleadoComision(Empleado): 
    def __init__(self, nombre, salario_base, ventas, porcentaje_comision): 
        super().__init__(nombre, salario_base) 
        self.ventas= ventas
        self.porcentaje_comision= porcentaje_comision

    def calcular_salario(self): 
        comision= self.ventas * self.porcentaje_comision / 100 
        return self.salario_base + comision 
    
class EmpleadoTemporal(Empleado): 
    def __init__(self, nombre, dias_trabajados, pago_por_dia, salario_base= None): 
        #salario_base no se necesita pero se mantiene por herencia: 
        super().__init__(nombre, salario_base)
        self.dias_trabajados= dias_trabajados
        self.pago_por_dia= pago_por_dia

    def calcular_salario(self):
        return self.dias_trabajados * self.pago_por_dia
    
def mostrar_salario(empleado): 
    print(f"Empleado: {empleado.nombre} -> Salario: ${empleado.calcular_salario():.2f}")

#Programa principal 
empleados = [ 
    Empleado("Orlando", 3000),
    EmpleadoComision("Marco", 2000, 50000, 5), 
    EmpleadoTemporal("Carlos", 20, 150)
]

print("=== Lista de empleados ===")

for emp in empleados: 
    mostrar_salario(emp)

print("=== Suma de salarios usando sobrecarga '+' ===")
suma= empleados[0] + empleados[1]
print(f"Orlando + Marco = ${suma:.2f}")

#suma_total= empleados[0] + empleados[1] + empleados[2]
#print(f"Suma total Orlando + Marco + Carlos = ${suma_total:.2f}")

File: RA 2.1/Actividad 5.cpp
#include <iostream>
#include <vector> 
#include <iomanip>
using namespace std; 

class Empleado { 
    protected: 
        string nombre;
        double salario_base; 

    public: 
        Empleado(string n, double sb): nombre(n), salario_base(sb) {}
        virtual ~Empleado() {}

        virtual double calcular_salario() {return salario_base;}
        string getNombre() const {return nombre;}

        double operator+(Empleado & otro) {
            return this->calcular_salario() + otro.calcular_salario();
        } 

        string operator<(Empleado & otro) {
            if (this->calcular_salario() < otro.calcular_salario()) {
                return otro.nombre+" tiene un mayor salario que "+this->nombre;
            }
            else if (otro.calcular_salario() < this->calcular_salario()) { 
                return this->nombre+" tiene un mayor salario que "+otro.nombre;
            }
            else { 
                return this->nombre+" y "+otro.nombre+" tienen el mismo salario";
            }
        }

};

class EmpleadoComision : public Empleado {
    double ventas;
    double porcentaje_comision; 

    public: 
        EmpleadoComision(string n, double sb, double v, double pcom): Empleado(n, sb), ventas(v), porcentaje_comision(pcom) {}

        double calcular_salario() override { 
            double comision= this->ventas * this->porcentaje_comision / 100;
            return this->salario_base + comision;
        }
};

class EmpleadoTemporal : public Empleado {
    double dias_trabajados;  
    double pago_por_dia;

    public: 
        EmpleadoTemporal(string n, double dt, double ppd, int sb= 0): Empleado(n, sb), dias_trabajados(dt), pago_por_dia(ppd) {}
        
        double calcular_salario() override { 
            return this->dias_trabajados * this->pago_por_dia;
        }
};

void mostrar_salario(Empleado* emp) {
    cout<<"Empleado: "<<emp->getNombre()
    <<fixed<<setprecision(2) 
    <<" -> Salario: $"<<emp->calcular_salario()<<endl;
}

int main() { 
    vector<Empleado*> empleados= {
        new Empleado("Orlando", 3000),
        new EmpleadoComision("Marco", 2000, 50000, 5),
        new EmpleadoTemporal("Carlos", 20, 150)
    };

    cout<<"=== Lista de empleados ==="<<endl;
    for (auto emp : empleados) {
        mostrar_salario(emp);
    }

    cout<<"=== Suma de salarios usando sobrecarga '+' ==="<<endl;
    double suma= *empleados[0] + *empleados[1]; 
    cout<<"Orlando + Marcos= $"<<suma<<endl;
    cout<<(*empleados[1] < *empleados[0])<<endl;

    for (auto emp : empleados) delete emp;

    return 0;
}